annotation/app/annotations/main.py

```diff
@@ -60,14 +66,35 @@ def row_to_dict(row) -> dict:
     }
 
 
+class NotConfiguredException(Exception):
+    pass
+
+
 def connect_s3(bucket_name: str) -> boto3.resource:
-    s3_resource = boto3.resource(
-        "s3",
-        endpoint_url=ENDPOINT_URL,
-        aws_access_key_id=AWS_ACCESS_KEY_ID,
-        aws_secret_access_key=AWS_SECRET_ACCESS_KEY,
-    )
+    boto3_config = {}
+    if S3_CREDENTIALS_PROVIDER == "minio":
+        logger_.info(f"S3_Credentials provider - {S3_CREDENTIALS_PROVIDER}")
+        boto3_config.update(
+            {
+                "aws_access_key_id": AWS_ACCESS_KEY_ID,
+                "aws_secret_access_key": AWS_SECRET_ACCESS_KEY,
+                "endpoint_url": ENDPOINT_URL,
+            }
+        )
+    elif S3_CREDENTIALS_PROVIDER == "aws_iam":
+        logger_.info(f"S3_Credentials provider - {S3_CREDENTIALS_PROVIDER}")
+        # No additional updates to config needed - boto3 uses env vars
+    else:
+        raise NotConfiguredException(
+            "s3 connection is not properly configured - "
+            "s3_credentials_provider is not set"
+        )
+    s3_resource = boto3.resource("s3", **boto3_config)
+
     try:
+        if S3_PREFIX:
+            bucket_name = f"{S3_PREFIX}-{bucket_name}"
```

Comment:
I think we shouldn't change a bucket name here. `connect_s3` takes `bucket_name` already. We hide this logic here, a bit confusing.

Reply:
It's one of the main functions, with which all s3_communications work.
And this prefix insertion is needed to check whether the bucket exists after connection.
If yes - everything's fine, if no - something with connection gone wrong

Reply:
But `connect_s3` clearly works with S3, it doesn't know about tenants and prefixes at all. Again, `connect_s3` takes `bucket_name`, not `tenant`. We either change it to `tenant` and do modifications inside (IMO it's a bad path), or simply pass only bucket name to work with, which is pretty straightforward. 
The issue we may face when you have only bucket name (and if we have prefix, it's already prefixed). You pass it to the `connect_s3` which modifies it again, at the end you get double prefixed bucket name.

Reply:
Agreed. I've done this switch to bucket_name with prefix in functions layer above